fragment/fragment/src/main/java/androidx/fragment/app/strictmode/FragmentStrictMode.java

```diff
@@ -0,0 +1,204 @@
+            private final Set<Flag> flags;
+            private OnViolationListener listener;
+
+            /** Create a Builder that detects nothing and has no violations. */
+            public Builder() {
+                flags = new HashSet<>();
+            }
+
+            /** Log detected violations to the system log. */
+            @NonNull
+            @SuppressLint("BuilderSetStyle")
+            public Builder penaltyLog() {
+                flags.add(Flag.PENALTY_LOG);
+                return this;
+            }
+
+            /**
+             * Crash the whole process on violation. This penalty runs at the end of all enabled
+             * penalties so you'll still get to see logging or other violations before the process
+             * dies.
+             */
+            @NonNull
+            @SuppressLint("BuilderSetStyle")
+            public Builder penaltyDeath() {
+                flags.add(Flag.PENALTY_DEATH);
+                return this;
+            }
+
+            /** Call #{@link OnViolationListener#onViolation} for every violation. */
+            @NonNull
+            @SuppressLint("BuilderSetStyle")
+            public Builder penaltyListener(@NonNull OnViolationListener listener) {
+                this.listener = listener;
+                return this;
+            }
+
+            /**
+             * Construct the Policy instance.
+             *
+             * <p>Note: if no penalties are enabled before calling <code>build</code>, {@link
+             * #penaltyLog} is implicitly set.
+             */
+            @NonNull
+            public Policy build() {
+                if (listener == null && !flags.contains(Flag.PENALTY_DEATH)) {
+                    penaltyLog();
+                }
+                return new Policy(flags, listener);
+            }
+        }
+    }
+
+    /** Returns the current default policy. */
+     @NonNull
+    public static Policy getDefaultPolicy() {
+        return defaultPolicy;
+    }
+
+    /**
+     * Sets the policy for what actions should be detected, as well as the penalty if such actions
+     * occur.
+     *
+     * @param policy the policy to put into place
+     */
+    public static void setDefaultPolicy(@NonNull Policy policy) {
+        defaultPolicy = policy;
+    }
+
+    private static Policy getNearestPolicy(@Nullable Fragment fragment) {
+        while (fragment != null) {
+            if (fragment.isAdded()) {
+                FragmentManager fragmentManager = fragment.getParentFragmentManager();
+                if (fragmentManager.getStrictModePolicy() != null) {
+                    return fragmentManager.getStrictModePolicy();
+                }
+            }
+            fragment = fragment.getParentFragment();
+        }
+        return defaultPolicy;
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.NONE)
+    static void onPolicyViolation(@NonNull Fragment fragment, @NonNull Violation violation) {
```

Comment:
I'm not particularly enthusiastic about this static call from anywhere as I think it precludes the host Handler / Executor discussion above.

Reply:
Depends on what we should run on the `Handler`/`Executor` I think. We could either

1. just trigger the listeners or
2. do the whole processing (logging, checking policy, listeners, ...)

on the `Executor`. In either case, I think we can easily switch get the right `Handler`, since we get the violating `Fragment` as parameter. Another option I can think of would be to make `FragmentStrictMode` a singleton and keep a reference to it in each `Fragment`, but I'm not sure if that's actually better.

Reply:
I'd be okay if the follow up CL we talked about had another version of this that took in the `Handler` did the synchronous / posted dispatch I mentioned in my other comment. That would give the right behavior without necessarily precluding this static method style.